src/libraries/Microsoft.Extensions.DependencyInjection/tests/DI.Tests/ServiceProviderContainerTests.cs

```diff
@@ -265,6 +267,128 @@ public void ScopeDispose_PreventsServiceResolution()
             Assert.NotNull(provider.CreateScope());
         }
 
+        [Fact]
+        public void GetService_ThenDisposeOnDifferentThread_ServiceDisposed()
+        {
+            var services = new ServiceCollection();
+            services.AddSingleton<Foo1>();
+            IServiceProvider sp = services.BuildServiceProvider();
+            var foo = sp.GetRequiredService<Foo1>();
+            Task.Run(() => (sp as IDisposable).Dispose()).Wait();
+            
+            Assert.True(foo.IsDisposed);
+        }
+
+        [Fact]
+        public void GetService_DisposeOnSameThread_Throws()
+        {
+            var services = new ServiceCollection();
+            services.AddSingleton<Foo2>();
+            IServiceProvider sp = services.BuildServiceProvider();
+            Assert.Throws<ObjectDisposedException>(() =>
+            {
+                // ctor disposes ServiceProvider
+                var foo = sp.GetRequiredService<Foo2>();
+            });
+        }
+
+        [Fact]
+        public void GetAsyncService_DisposeAsyncOnSameThread_ThrowsAndDoesNotHang()
+        {
+            // Arrange
+            var services = new ServiceCollection();
+            services.AddSingleton<FooAsyncDisposable>();
+            var sp = services.BuildServiceProvider();
+            bool success = Task.Run(() =>
+            {
+                SingleThreadedSynchronizationContext.Run(() =>
+                {
+                    Task.Factory.StartNew(() => 
+                    {
+                        // Act
+                        Assert.Throws<ObjectDisposedException>(() =>
+                        {
+                            // ctor disposes ServiceProvider
+                            var foo = sp.GetRequiredService<FooAsyncDisposable>();
+                        });
+                    },
+                    CancellationToken.None, 
+                    TaskCreationOptions.None,
+                    TaskScheduler.FromCurrentSynchronizationContext()
+                    ).Wait();
+                });
+            }).Wait(TimeSpan.FromSeconds(10));
+
+            Assert.True(success);
+        }
+
+        private class Foo1 : IDisposable
+        {
+            public Foo1()
+            {
+                Thread.Sleep(5000);
+            }
+            public bool IsDisposed { get; private set; }
+
+            public void Dispose()
+            {
+                IsDisposed = true;
+            }
+        }
+
+        private class Foo2 : IDisposable
```

Comment:
I would name this class appropriately. `private class DisposeServiceProviderInCtor`